File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/OxygenSaturationRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.OxygenSaturationRecord
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.OxygenSaturationRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [OxygenSaturationRecord] to a [OxygenSaturationRecordDto].
 */
internal fun OxygenSaturationRecord.toDto(): OxygenSaturationRecordDto = OxygenSaturationRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    percentage = percentage.toDto(),
)

/**
 * Converts a [OxygenSaturationRecordDto] to a Health Connect [OxygenSaturationRecord].
 */
internal fun OxygenSaturationRecordDto.toHealthConnect(): OxygenSaturationRecord =
    OxygenSaturationRecord(
        percentage = percentage.toHealthConnect(),
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(),
    )

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/oxygen_saturation_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/measurement_unit_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:meta/meta.dart' show internal;

/// Mapper extensions for [OxygenSaturationRecord].
@sinceV1_3_0
@internal
extension OxygenSaturationRecordToDto on OxygenSaturationRecord {
  /// Converts [OxygenSaturationRecord] to its [OxygenSaturationRecordDto].
  OxygenSaturationRecordDto toDto() {
    return OxygenSaturationRecordDto(
      id: id.toDto(),
      time: time.toUtc().millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      percentage: percentage.toDto(),
    );
  }
}

/// Mapper extensions for [OxygenSaturationRecordDto].
@sinceV1_3_0
@internal
extension OxygenSaturationRecordDtoToDomain on OxygenSaturationRecordDto {
  /// Converts [OxygenSaturationRecordDto] to its domain representation.
  OxygenSaturationRecord toDomain() {
    return OxygenSaturationRecord(
      id: HealthRecordId(id ?? HealthRecordId.none.value),
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true).toLocal(),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      percentage: percentage.toDomain(),
    );
  }
}
